### ISSUE TITLE
Negative Loss after certain number of epochs

### ISSUE THREAD

**Issue by rtaiello** [facebookresearch/CrypTen#347]
Hi, 

I'm training a logistic regression using the following code, and at certain point the loss stop to decrease and it gets negative values:
```
CLIENT_0 = 0
CLIENT_1 = 1
dummy_input = torch.empty(1, 1, 784)

@mpc.run_multiprocess(world_size=2)
def run_encrypted_training():
    global_parameters = torch.zeros(NUM_ROUNDS,NUM_PARAMETERS)
    global_gradients = torch.zeros(NUM_ROUNDS,NUM_PARAMETERS)
    # Load data:
    X_client_0_enc = crypten.load_from_party('/tmp/client_0_train.pth', src=CLIENT_0)
    X_client_1_enc = crypten.load_from_party('/tmp/client_1_train.pth', src=CLIENT_1)
    
    y_client_0_enc = crypten.load_from_party('/tmp/client_0_train_labels.pth', src=CLIENT_0)
    y_client_1_enc = crypten.load_from_party('/tmp/client_1_train_labels.pth', src=CLIENT_1)
    
    X_combined_enc = crypten.cat([X_client_0_enc, X_client_1_enc], dim=0)
    
    y_combined_enc = crypten.cat([y_client_0_enc, y_client_1_enc], dim=0)
     
    pytorch_model = LogisticRegression()
    pytorch_model.load_state_dict(inizitial_parameters)
    model = crypten.nn.from_pytorch(pytorch_model, dummy_input)
    # Encrypt the model, namely the model is split between the two clients
    model.encrypt()
    
    model.train()
  
    # Define a loss function using crypten
    loss = crypten.nn.CrossEntropyLoss()

    # Define training parameters
    learning_rate = 1e-3
    num_epochs = 1
    batch_size = 100
    num_batches = X_combined_enc.size(0) // batch_size
        
    rank = comm.get().get_rank()
    for round in trange(NUM_ROUNDS, desc="Round"):
        for i in range(num_epochs):             
            for batch in range(num_batches):
                start, end = batch * batch_size, (batch + 1) * batch_size
                                        
                X_batch = X_combined_enc[start:end].reshape(batch_size,-1)
                y_batch = y_combined_enc[start:end].reshape(batch_size,-1)
                # perform forward pass:

                output = model(X_batch)

                loss_value = loss(output, y_batch)
                
                model.zero_grad()

                loss_value.backward()
                # Taking the gradient at the end of the current ROUND
                if batch == num_batches-1 and i == num_epochs-1:
                    crypten.print("Get the gradients")
                    global_gradients[round] = torch.cat([p.grad.get_plain_text().flatten() for p in model.parameters()])
                # crypten.print([p.grad.get_plain_text() for p in model.parameters()])
                # print([p.grad.get_plain_text() for p in model.parameters()])


                model.update_parameters(learning_rate)
                
                batch_loss = loss_value.get_plain_text()
        crypten.print(batch_loss)
        # for each round we saved the global model's parameters
        crypten.print("Get the parameters")
   ```
   The output were tensor() is the loss at each round
   
 ```
 Get the gradients
tensor(1.2044)
Get the parameters
Round:   1%|          | 1/100 [04:14<7:00:38, 254.94s/it]Get the gradients
tensor(0.9808)
Get the parameters
Round:   2%|▏         | 2/100 [08:29<6:55:44, 254.54s/it]Get the gradients
tensor(0.8806)
Get the parameters
Round:   3%|▎         | 3/100 [12:43<6:51:23, 254.47s/it]Get the gradients
tensor(0.8219)
Get the parameters
Round:   4%|▍         | 4/100 [16:57<6:46:39, 254.16s/it]Get the gradients
tensor(0.7821)
Get the parameters
Round:   5%|▌         | 5/100 [21:11<6:42:27, 254.18s/it]Get the gradients
tensor(0.7533)
Get the parameters
Round:   6%|▌         | 6/100 [25:26<6:38:32, 254.39s/it]Get the gradients
tensor(0.7314)
Get the parameters
Round:   7%|▋         | 7/100 [29:40<6:34:10, 254.31s/it]Get the gradients
tensor(0.7136)
Get the parameters
Round:   8%|▊         | 8/100 [33:54<6:29:42, 254.15s/it]Get the gradients
tensor(0.6992)
Get the parameters
Round:   9%|▉         | 9/100 [38:07<6:25:01, 253.87s/it]Get the gradients
tensor(0.6868)
Get the parameters
Round:  10%|█         | 10/100 [42:21<6:20:59, 253.99s/it]Get the gradients
tensor(0.6758)
Get the parameters
Round:  11%|█         | 11/100 [46:35<6:16:33, 253.86s/it]Get the gradients
tensor(0.6659)
Get the parameters
Round:  12%|█▏        | 12/100 [50:48<6:12:00, 253.64s/it]Get the gradients
tensor(0.6578)
Get the parameters
Round:  13%|█▎        | 13/100 [55:02<6:07:55, 253.75s/it]Get the gradients
tensor(0.6501)
Get the parameters
Round:  14%|█▍        | 14/100 [59:17<6:04:03, 253.99s/it]Get the gradients
tensor(0.6435)
Get the parameters
Round:  15%|█▌        | 15/100 [1:03:33<6:00:42, 254.61s/it]Get the gradients
tensor(0.6371)
Get the parameters
Round:  16%|█▌        | 16/100 [1:07:47<5:56:26, 254.60s/it]Get the gradients
tensor(0.6315)
Get the parameters
Round:  17%|█▋        | 17/100 [1:12:01<5:51:59, 254.45s/it]Get the gradients
tensor(0.6263)
Get the parameters
Round:  18%|█▊        | 18/100 [1:16:16<5:47:50, 254.51s/it]Get the gradients
tensor(-2.7052e+08)
Get the parameters
Round:  19%|█▉        | 19/100 [1:20:30<5:43:32, 254.47s/it]Get the gradients
tensor(84239000.)
Get the parameters
Round:  20%|██        | 20/100 [1:24:44<5:38:53, 254.17s/it]Get the gradients
tensor(50677880.)
Get the parameters
Round:  21%|██        | 21/100 [1:28:58<5:34:33, 254.10s/it]Get the gradients
tensor(25419306.)
Get the parameters
Round:  22%|██▏       | 22/100 [1:33:12<5:30:20, 254.11s/it]Get the gradients
tensor(3.0287e+08)
Get the parameters
Round:  23%|██▎       | 23/100 [1:37:25<5:25:55, 253.97s/it]Get the gradients
tensor(-1.4610e+08)
Get the parameters
Round:  24%|██▍       | 24/100 [1:41:39<5:21:28, 253.79s/it]Get the gradients
tensor(-2.3147e+08)
Get the parameters
Round:  25%|██▌       | 25/100 [1:45:53<5:17:15, 253.80s/it]
Thanks in advance!

**Comment by aeasy371** [comment]
Hi, I am also having this issue, did you fix it?

**Comment by knottb** [comment]
This behavior is likely due to an artifact of the encryption mechanism that has a causes the result of a multiplication / division operation to be undefined with very low probability. If this happens, it can propagate overflow values to the rest of your training. Because of this, it is usually the best option (if possible) to monitor the loss function, and restart from a checkpoint if you see exploding loss values.

Because this happens with very low probability, this artifact tends to appear only in computations that require a lot of operations (e.g. large models and/or large data). Please let me know if this fixes your problem.

**Comment by aeasy371** [comment]
I concur that it is likely an overflow, thank you for the input. I am currently training with a learning rate roughly 100x smaller than previously and it seems that the gradient updates are much more stable, but slower by quite a margin on training. The model I am training is as follows:


```
class FeatureSqueezeModel(crypten.nn.Module):
	def __init__(self):
		super(FeatureSqueezeModel, self).__init__()
		self.squeeze_layer = crypten.nn.Linear(768, 480)  # Reduce feature size from 768 → 480
	def forward(self, x):
		return self.squeeze_layer(x)  # Output shape: [batch_size, 1, 200, 480]
class SecureClassifierModel(crypten.nn.Module):
	def __init__(self):
		super(SecureClassifierModel, self).__init__()

		self.feature_squeeze = FeatureSqueezeModel()  # Add squeeze step

		self.relu1 = crypten.nn.ReLU()
		
		# Convolutional layer with kernel size (200,1)
		self.conv = crypten.nn.Conv2d(in_channels=1, out_channels=32, kernel_size=(200, 1), stride=(1, 1), padding=(0, 0))
		self.relu2 = crypten.nn.ReLU()
		# self.batchnorm2d = crypten.nn.BatchNorm2d(32)
		# Flatten output
		crypten.nn.Flatten(),

		# Fully connected layers
		self.fc1 = crypten.nn.Linear(32 * 480, 256)
		self.relu3 = crypten.nn.ReLU()
		# self.batchnorm1d = crypten.nn.BatchNorm1d(256)
		self.fc2 = crypten.nn.Linear(256, 2)  # Binary classification output

	def forward(self, x):
		batch_size = x.shape[0]  # Get batch size

		x = self.feature_squeeze(x)  # Reduce feature dim from 768 → 480
		x = x.view(batch_size, 1, 200, 480)  # Reshape for Conv2D

		x = self.conv(x)  # Apply Conv2D layer
		x = self.relu2(x)
		# x = self.batchnorm2d(x)

		x = x.view(batch_size, -1)  # Flatten for the next FC layers
		x = self.fc1(x)
		x = self.relu3(x)
		# x = self.batchnorm1d(x)

		x = self.fc2(x)  # Final output layer
		return x
```

which I have not thought of as "large", I understand that there is some issue with batchnorm in CrypTen so I have removed it from the code. If I am training in a two party system with an ecrypted model and input data, do you have any guidance on the best method to save a model as a checkpoint and load it back in correctly? I have had trouble finding accurate training example code with a multiparty system.
